thunder/core/trace_interpreter.py

```diff
@@ -74,3 +77,129 @@ def write(v: VariableInterface | Any, val: Any, allow_duplicates=False) -> None:
         return tree_map(read, trace.output), env
 
     return tree_map(read, trace.output)
+
+
+def interpret_trace_to_trace(trace, *args, symbol_mapper=None, with_env=False, **kwargs):
+    """Interpret a trace.
+
+    Args:
+        trace: trace to interpret
+        *args: arguments to interpret the trace with
+        symbol_mapper: function that redirects the evaluation of a BoundSymbol to a different function
+        with_env: whether to return the environment after interpreting the trace. Environment is a dictionary
+            that maps VariableInterface objects to their values.
+        **kwargs: keyword arguments to interpret the trace with
+
+    Returns:
+        result of interpreting the trace, optionally with the environment that saves all intermediate values
+    """
+    env = {}
+
+    def read(x: VariableInterface | Any) -> Any:
+        if isinstance(x, VariableInterface):
+            return env[x.name]
+        else:
+            return x
+
+    def write(v: VariableInterface | Any, val: Any, allow_duplicates=False) -> None:
+        if not isinstance(v, VariableInterface):
+            return
+        # Duplicates are allowed and overwritten
+        if v.name in env:
+            if allow_duplicates:
+                return
+            raise ValueError(f"Variable {v.name} is being overwritten this is not allowed")
+        env[v.name] = val
+
+    def add_to_swap_map(old, new):
+        if isinstance(old, ProxyInterface):
+            if isinstance(new, ProxyInterface) and variableify(new) in env:
+                # the new isn't new, but something returned the input
+                # this means we need to map the old to the new
+                old, new = new, old
+            elif isinstance(old, TensorProxyInterface):
+                # should we have a fix shapes pass? the sharding
+                # (FSDP, tensor parallel) transforms do "break" shape metadata
+                new_trace.names.remove(old.name)  # taken by the .replace proxy
+                if isinstance(new, VJPDual):
+                    old = old.replace(shape=new.primal.shape)
+                else:
+                    old = old.replace(shape=new.shape)
+
+            if isinstance(new, VJPDual):
+                swap_map[variableify(new.primal)] = old
+                new.primal = old
+            else:
+                assert isinstance(new, ProxyInterface), (old, new)
+                swap_map[variableify(new)] = old
+
+    def do_swap(v):
+        if isinstance(v, VJPDual):
+            v.primal = tree_map(do_swap, v.primal)
+            v.residuals = tree_map(do_swap, v.residuals)
+            return v
+        if not isinstance(v, ProxyInterface):
+            return v
+        return swap_map.get(variableify(v), v)
+
+    new_trace = from_trace(trace)
+    with tracectx(new_trace):
+        swap_map = {}
+
+        safe_map_flat(add_to_swap_map, list(trace.args), list(args))
```

Comment:
should we check the new args and kwargs are proper replacements? (same number of args, same keys)

(not strictly needed, can go as is)

Reply:
I think it would error, that's the "safe"...